bootstrap/updater.go

```diff
@@ -0,0 +1,407 @@
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+const (
+	DefaultURL = "http://localhost:3000/bootstrap"
+	DirName    = "bootstrap"
+
+	httpTimeout   = 5 * time.Second
+	notifyTimeout = time.Second
+	schemaFile    = "schema.json"
+	format        = "2006-01-02T15-04-05"
+)
+
+var (
+	ErrEpochOutOfOrder = errors.New("epoch out of order")
+	ErrWrongVersion    = errors.New("wrong schema version")
+	ErrInvalidBeacon   = errors.New("invalid beacon")
+)
+
+type Config struct {
+	URL     string `mapstructure:"bootstrap-uri"`
+	Version string `mapstructure:"bootstrap-version"`
+
+	DataDir   string
+	Interval  time.Duration
+	NumToKeep int
+}
+
+func DefaultConfig() Config {
+	return Config{
+		URL:       DefaultURL,
+		Version:   "https://spacemesh.io/bootstrap.schema.json.1.0",
+		DataDir:   os.TempDir(),
+		Interval:  30 * time.Second,
+		NumToKeep: 10,
+	}
+}
+
+type Updater struct {
+	cfg    Config
+	logger log.Log
+	fs     afero.Fs
+	client *http.Client
+	once   sync.Once
+	eg     errgroup.Group
+
+	mu          sync.Mutex
+	subscribers []chan *VerifiedUpdate
+	latest      *VerifiedUpdate
+}
+
+type Opt func(*Updater)
+
+func WithConfig(cfg Config) Opt {
+	return func(u *Updater) {
+		u.cfg = cfg
+	}
+}
+
+func WithLogger(logger log.Log) Opt {
+	return func(u *Updater) {
+		u.logger = logger
+	}
+}
+
+func WithFilesystem(fs afero.Fs) Opt {
+	return func(u *Updater) {
+		u.fs = fs
+	}
+}
+
+func WithHttpClient(c *http.Client) Opt {
+	return func(u *Updater) {
+		u.client = c
+	}
+}
+
+func New(opts ...Opt) *Updater {
+	u := &Updater{
+		cfg:    DefaultConfig(),
+		logger: log.NewNop(),
+		fs:     afero.NewOsFs(),
+		client: &http.Client{},
+	}
+	for _, opt := range opts {
+		opt(u)
+	}
+	return u
+}
+
+func (u *Updater) Subscribe() chan *VerifiedUpdate {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	ch := make(chan *VerifiedUpdate, 10)
+	u.subscribers = append(u.subscribers, ch)
+	return ch
+}
+
+func (u *Updater) Load(ctx context.Context) error {
+	verified, err := load(u.fs, u.cfg)
+	if err != nil {
+		return err
+	}
+	if verified != nil {
+		if err = u.updateAndNotify(ctx, verified); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (u *Updater) Start(ctx context.Context) {
+	u.once.Do(func() {
+		u.eg.Go(func() error {
+			if err := u.Load(ctx); err != nil {
+				return err
+			}
+			wait := time.Duration(0)
+			for {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-time.After(wait):
+					ctx := log.WithNewSessionID(ctx)
+					if err := u.DoIt(ctx); err != nil {
+						u.logger.With().Error("failed to get bootstrap update", log.Err(err))
+					}
+				}
+				wait = u.cfg.Interval
+			}
+		})
+	})
+}
+
+func (u *Updater) Close() {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	for _, ch := range u.subscribers {
+		close(ch)
+	}
+	_ = u.eg.Wait()
+}
+
+func (u *Updater) latestUpdateId() uint32 {
+	u.mu.Lock()
+	defer u.mu.Unlock()
```

Comment:
I don't think we need to lock here. `u.latest` seems like its only accessed by the single internal go routine.

Reply:
i don't think one should conditionally lock mutable data. 
the only data i didn't protect are those that are assigned in `New` and never get changed in the object's life cycle.

this module is not performance critical. for those that are changed outside of `New`, i think it's safer/future-proof to protect always.

Reply:
Well, we disagree on this point.

I find locking of objects that do not need to be synchronised makes the code much harder to understand, since it implies to me that those objects are going to be accessed by multiple go-routines.

I'm also not convinced that it makes the code any safer since adding locked sections also raises the risk of deadlocks.

I think if we actually followed the approach of locking over all mutable fields the code-base would become exceedingly complex.

Reply:
ok. we disagreed.